### PR TITLE
ETH Instant Withdraw

### PR DESCRIPTION
**Pull request by zhoujia6139**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/interfaces/IPoolInstantWithdraw.sol [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        uint16 referralCode
+    ) external;
+
+    /**
+     * @notice swap a term loan collateral with the loan's borrow asset,
+     * the amount user need to pay is calculated by the present value of the collateral
+     * @param loanId The id for the specified loan
+     * @param receiver The address to receive the collateral asset
+     **/
+    function swapLoanCollateral(uint256 loanId, address receiver) external;
```
Comment (WalidOfNow) [inline contracts/interfaces/IPoolInstantWithdraw.sol:157]: why not just call it `repayLoan`?

File: contracts/protocol/pool/PoolInstantWithdraw.sol [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+    /// @inheritdoc IPoolInstantWithdraw
+    function createLoan(
+        address collateralAsset,
+        uint256 collateralTokenId,
+        address borrowAsset,
+        uint256 borrowAmount,
+        uint16 referralCode
+    ) external nonReentrant {
+        DataTypes.PoolStorage storage ps = poolStorage();
+        _validateBorrowAsset(ps, collateralAsset, borrowAsset);
```
Comment (WalidOfNow) [inline contracts/protocol/pool/PoolInstantWithdraw.sol:253]: why this is not in validation logic?

File: contracts/protocol/pool/PoolInstantWithdraw.sol [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+        // check if borrow amount is valid
+        {
+            // fetch present value and discount rate from Oracle
+            (presentValue, discountRate) = IInstantNFTOracle(WITHDRAW_ORACLE)
+                .getPresentValueAndDiscountRate(
+                    collateralTokenId,
+                    nextBorrowRate
+                );
+
+            uint256 borrowAssetAmount = _calculateAssetAmountByETHValue(
```
Comment (WalidOfNow) [inline contracts/protocol/pool/PoolInstantWithdraw.sol:298]: how about `_calculatePresentValueInBorrowAsset`

Comment (WalidOfNow) [reply contracts/protocol/pool/PoolInstantWithdraw.sol:298]: and the var `presentValueInBorrowAsset`

File: contracts/protocol/pool/PoolInstantWithdraw.sol [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+        address collateralAsset = loan.collateralAsset;
+        uint256 collateralTokenId = uint256(loan.collateralTokenId);
+        address borrowAsset = loan.borrowAsset;
+        //here oracle need to guarantee presentValue > debtValue.
+        uint256 presentValue = IInstantNFTOracle(WITHDRAW_ORACLE)
+            .getPresentValueByDiscountRate(
+                collateralTokenId,
+                loan.discountRate
+            );
+        uint256 borrowAssetAmountNeeded = _calculateAssetAmountByETHValue(
```
Comment (WalidOfNow) [inline contracts/protocol/pool/PoolInstantWithdraw.sol:404]: same as above

File: contracts/protocol/pool/PoolInstantWithdraw.sol [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+        );
+
+        // update borrow asset state
+        DataTypes.ReserveCache memory reserveCache;
+        DataTypes.ReserveData storage reserve = ps._reserves[borrowAsset];
+        reserveCache = reserve.cache();
+        reserve.updateState(reserveCache);
+
+        // repay borrow asset debt and update interest rate
+        uint256 totalDebt = _calculateTotalStableDebt(
```
Comment (WalidOfNow) [inline contracts/protocol/pool/PoolInstantWithdraw.sol:416]: how about `_calculateLoanStableDebt`
and the var `loanDebt` or `loanStableDebt`

File: contracts/protocol/pool/PoolInstantWithdraw.sol [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+        ILoanVault LoanVault = ILoanVault(VAULT_CONTRACT);
+        LoanVault.settleCollateral(collateralAsset, collateralTokenId);
+
+        // repay borrow asset debt and update interest rate
+        uint256 totalDebt = _calculateTotalStableDebt(
+            loan.borrowAmount,
+            loan.stableRate,
+            loan.startTime
+        );
+        uint256 vaultBalance = IERC20(borrowAsset).balanceOf(VAULT_CONTRACT);
```
Comment (WalidOfNow) [inline contracts/protocol/pool/PoolInstantWithdraw.sol:488]: `vaultBorrowAssetBalance`

File: contracts/protocol/tokenization/ATokenStableDebtToken.sol [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import {RebaseStableDebtToken} from "./RebaseStableDebtToken.sol";
+import {WadRayMath} from "../libraries/math/WadRayMath.sol";
+import {IAToken} from "../../interfaces/IAToken.sol";
+
+/**
+ * @title aToken Rebasing Debt Token
+ *
+ * @notice Implementation of the interest bearing token for the ParaSpace protocol
+ */
+contract ATokenStableDebtToken is RebaseStableDebtToken {
```
[the remark this answers is not in the document]
Comment (zhoujia6139) [reply contracts/protocol/tokenization/ATokenStableDebtToken.sol:14]: ATokenStableDebtToken is a debt token. It's for aWETH or astETH or awstETH.

**Pull request closed without merging**